lib/src/client.dart

```diff
@@ -140,3 +141,48 @@ abstract class Client {
   /// do so can cause the Dart process to hang.
   void close();
 }
+
+const _clientToken = Object();
+
+/// Return the [Client] for the current [Zone], if one has been set.
+///
+/// NOTE: This function is explicitly hidden from the public API.
+Client? get zoneClient => Zone.current[_clientToken] != null
+    ? Zone.current[_clientToken] as Client
+    : null;
+
+/// Runs [body] in its own [Zone] with [client] set as the default [Client].
+///
+/// For example:
+///
+/// ```
+/// class MyHttpClient extends http.BaseClient {
+///   @override
+///   Future<http.StreamedResponse> send(http.BaseRequest request) {
+///     // your implementation here
+///   }
+/// }
+///
+/// void main() {
+///  late Client client;
+///  if (Platform.isAndroid) {
+///    client = MyHttpClient();
+///  } else {
+///    client = Client();  // Use the default client.
+///  }
+///  runWithClient(myFunction, () => client);
+/// }
+///
+/// myFunction() {
+///   // Uses the `Client` configured in `main`.
+///   final response = await get(Uri.parse("https://www.example.com/"));
+///   final client = Client();
+/// }
+/// ```
+///
+/// The [client] influences functions (e.g. `get`) and the [Client.new] factory.
+R runWithClient<R>(R Function() body, Client Function() client,
```

Comment:
*Consider* putting the `body` argument last. It reads better as:
```dart
runWIthClient(createClient, () {
   the body 
   of code
});
```
than putting the (usually small) client factory after the bigger body.

(Try it, and compare the readability of the "nested" tests.)

Reply:
I don't think it's necessarily a good idea, but we could consider a named argument for `createClient` and then the caller can put it in either order.

Reply:
I considered named parameters, but I'm generally not fond of required named parameters. Because I hate writing the name.

That said, in this case both arguments are likely to be function literals, so it might be confusing to both authors and readers which is which. That is one of the situations where the style guide suggests naming parameters.

I disagree with the ordering. You want the small method body *first*. If it's very short, like a single line, you can see past it and find the other function. A large first function literal makes it very hard to find the second one.

Being consistent has value, so let's just go with body last. If someone disagrees, they can write their own two-parameter wrapper function.

Reply:
> Being consistent has value, so let's just go with body last.

Do you mean we should be consistent with [`HttpOverrides.runWithHttpOverrides`](https://api.flutter.dev/flutter/dart-io/HttpOverrides/runWithHttpOverrides.html)? In that case it should be `body` first.